### PR TITLE
Addition of a CONTRIBUTING.md file to Quail

### PR DESCRIPTION
**Pull request by bornhoft**
This pull request adds a contributing file to Quail which gives some basic highlights of the code, a brief style guide, some naming conventions, and most importantly an example of interfacing with the abstract base classes often used in Quail.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1 +1,172 @@
-This is a place holder
+# Contributing to Quail
+
+Thanks for your interest in contributing! We are excited to have you join our team!
+
+The following is a set of guidelines for contributing to Quail [Ihme Group Organization](https://github.com/IhmeGroup/quail) on GitHub. Feel free to propose modifications to this document via a pull request as we value others contributions.
```
Comment (ericjching) [inline CONTRIBUTING.md:5]: others' instead of others

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1 +1,172 @@
...
+
+[Branch naming conventions](#branch-naming-conventions)
+## Getting Started
+
+Quail is a lightweight discontinuous Galerkin code written with the intention of being useful for teaching and prototyping. Quail is made up of a series of Python packages. These are all located in the `src` directory. They include the following:
+
+* [solver](https://github.com/IhmeGroup/quail/tree/main/src/solver) - which contains modules for various solvers (DG and ADER-DG). This is where the primary `solve` function is located which governs the loop around which iterations occur.
+* [meshing](https://github.com/IhmeGroup/quail/tree/main/src/meshing) - which contains modules for mesh-related classes and functions. Mesh generation tools are includedhere. For additional information on mesh generation, node ordering, and face ordering.
+* [numerics](https://github.com/IhmeGroup/quail/tree/main/src/numerics) - which contains modules for basis functions, quadrature, limiters, and time stepping.
+* [physics](https://github.com/IhmeGroup/quail/tree/main/src/physics) - which contains modules for various equation sets and related analytic functions, boundary conditions,and numerical fluxes.
```
Comment (ericjching) [inline CONTRIBUTING.md:23]: analytical instead of analytic (analytic function means something else in math)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1 +1,172 @@
...
+
+[Style Guide](#style-guide)
+
+[Branch naming conventions](#branch-naming-conventions)
+## Getting Started
+
+Quail is a lightweight discontinuous Galerkin code written with the intention of being useful for teaching and prototyping. Quail is made up of a series of Python packages. These are all located in the `src` directory. They include the following:
+
+* [solver](https://github.com/IhmeGroup/quail/tree/main/src/solver) - which contains modules for various solvers (DG and ADER-DG). This is where the primary `solve` function is located which governs the loop around which iterations occur.
+* [meshing](https://github.com/IhmeGroup/quail/tree/main/src/meshing) - which contains modules for mesh-related classes and functions. Mesh generation tools are includedhere. For additional information on mesh generation, node ordering, and face ordering.
```
Comment (ericjching) [inline CONTRIBUTING.md:21]: space between "included" and "here"

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1 +1,172 @@
...
+  * [Interfaces](#neat-but-how-do-i-interface-with-quails-packages)
+
+[Style Guide](#style-guide)
+
+[Branch naming conventions](#branch-naming-conventions)
+## Getting Started
+
+Quail is a lightweight discontinuous Galerkin code written with the intention of being useful for teaching and prototyping. Quail is made up of a series of Python packages. These are all located in the `src` directory. They include the following:
+
+* [solver](https://github.com/IhmeGroup/quail/tree/main/src/solver) - which contains modules for various solvers (DG and ADER-DG). This is where the primary `solve` function is located which governs the loop around which iterations occur.
```
Comment (ericjching) [inline CONTRIBUTING.md:20]: "governs the loop around which iterations occur" sounds a little awkward

maybe something simpler like "advances the solution in time"

File: CONTRIBUTING.md (hunk 3 above)
Comment (ericjching) [inline CONTRIBUTING.md:21]: Last sentence isn't a full sentence

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1 +1,172 @@
...
+    @property
+    @abstractmethod
+    def NEEDED_PROPERTY(self):
+      '''
+      This property is needed by every instantiation of this class.
+      '''
+    @abstractmethod
+    def need_this_function(self, args**):
+      '''
+      A function that everychild class of this type needs.
```
Comment (ericjching) [inline CONTRIBUTING.md:58]: space between "every" and "child"

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1 +1,172 @@
...
+      '''
+  ```
+We utilize `@abstractmethod` to define attributes that must be overwritten by any derived class. This ensures that the user defines the minimally required members of the class. Lets imagine that we wish add a new feature called `NeatFeature` that builds on the `NeatFeatureBase` class.
```
Comment (ericjching) [inline CONTRIBUTING.md:61]: Let's instead of Lets

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1 +1,172 @@
...
+  class NeatFeature(NeatFeatureBase):
+    '''
+    This neat feature inherits from the NeatFeatureBase class.
+    '''
+    # We define our abstract properties
+    NEEDED_PROPERTY = "Clearly this is needed"
+    # We define our abstract methods
+    def need_this_function(self, args**):
+      '''
+      How the new feature used this function.
```
Comment (ericjching) [inline CONTRIBUTING.md:73]: "uses" instead of "used"?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1 +1,172 @@
...
+  ```
+## Branch naming conventions
+
+Quail uses the following branch naming conventions:
+`<branch_type_name>/<branch_name>`
+There are multiple branch types that are standard in Quail. These include the following:
+* `bug`: these branches are designed for identifying, isolating, and fixing bugs that may exist throughout Quail. 
+* `feature`: these branches are designed for the addition of new features.
+* `test`: branches used for the specific addition of unit tests (tests can also be added throughout feature dev in the appropriate branch).
+* `case`: branches that specifically create new cases for Quail to solve. These can incorporate more case specific ICs, BCs, and exact solutions.
```
Comment (ericjching) [inline CONTRIBUTING.md:171]: "case-specific" instead of "case specific"

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1 +1,172 @@
...
+  	'''
+  ```
+## Branch naming conventions
+
+Quail uses the following branch naming conventions:
+`<branch_type_name>/<branch_name>`
+There are multiple branch types that are standard in Quail. These include the following:
+* `bug`: these branches are designed for identifying, isolating, and fixing bugs that may exist throughout Quail. 
+* `feature`: these branches are designed for the addition of new features.
+* `test`: branches used for the specific addition of unit tests (tests can also be added throughout feature dev in the appropriate branch).
```
Comment (ericjching) [inline CONTRIBUTING.md:170]: the wording inside the parentheses is kind of confusing

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1 +1,172 @@
...
+	return neat_feature
+  ```
+  
+These interfaces allow the user to focus on the meat of their new feature while not having too much overhead in making the feature available to the users.
+
+## Style Guide
+
+#### Some general guidelines to follow are:
+* Use tabs with a length of four spaces
+* Indent continued linestwice
```
Comment (ericjching) [inline CONTRIBUTING.md:124]: space between "lines" and "twice"

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1 +1,172 @@
...
+	'''
+	feature_name = params["NeatFeature"]
+	if NeatFeatureType[feature_name] == NeatFeatureType.NeatFeature:
+		neat_feature = neat_feature_defs.NeatFeature()
+	else:
+		raise NotImplementedError("Feature not supported")
+	return neat_feature
+  ```
+  
+These interfaces allow the user to focus on the meat of their new feature while not having too much overhead in making the feature available to the users.
```
Comment (ericjching) [inline CONTRIBUTING.md:118]: "not having too much overhead in" seems a bit wordy

maybe "simplifying the process of"

**Review by ericjching — APPROVED** [review]
Looks really good and definitely gives new users a good starting point for adding new features! My comments are mostly just on wording, typos, etc.

File: CONTRIBUTING.md (hunk 3 above)
Comment (bornhoft) [reply CONTRIBUTING.md:21]: Added a reference to the documentation and made this a sentence.

File: CONTRIBUTING.md (hunk 9 above)
Comment (bornhoft) [reply CONTRIBUTING.md:170]: agreed. I have removed this. It is trying to say that a test can be added in any kind of branch and that we don't have to make specific test only branches, but I think this is not so necessary.

**Review by alilasemi — APPROVED** [review]
Looks great!
